### PR TITLE
[NVPTX] Add patterns for fma.relu.{f16|f16x2|bf16|bf16x2}

### PR DESCRIPTION
**Pull request by hdelan**
Add patterns to lower `fmaxnum(fma(a, b, c), 0)` to `fma.rn{.ftz}.relu` for `f16`, `f16x2`, `bf16`, `bf16x2` types, when `nnan` is used.

`fma_relu` honours `NaN`, so the substitution is only made if the `fma` is `nnan`, since `fmaxnum` returns the non NaN argument when passed a NaN value.

This patch also removes some `bf16` ftz instructions since `FTZ` is not supported with the `bf16` type, according to the PTX ISA docs.

### CODE REVIEW

**Comment by hdelan** [comment]
Ping @ldrumm @frasercrmck

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (justinfargnoli) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3927]: Do we need the `Requires<...>` on the instruction too?

**Review by justinfargnoli — APPROVED** [review]
Overall, LGTM! 

Please wait for @AlexMaclean's review though as he's more familiar with `NVPTXInstrInfo.td` than I am.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3927]: It's only being used by the anonymous patterns below, which have the necessary `Requires`. I don't think we need to introduce extra noise by repeating them here.

File: llvm/test/CodeGen/NVPTX/fma-relu.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/fma-relu.ll:36]: It would be good to add a couple more test runs:
 - one with w/o mul/add -> fma contraction to make sure we do not use `fma.rn.relu` unintentionally.
 - one targeting older GPUs to make sure we do not emit `fma.rn.relu` there.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3927]: I think applying constraint to the instruction itself is the right thing to do. We do not want them to be emitted unintentionally, even if we do not do it now.

I do not know whether the constraint propagates to the pattern, but I think it may, so applying it here should do the job. It's easy enough to test by running the tests while targeting an older GPU.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3938]: Constraints look wrong. According to https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions-fma

```
PTX ISA Notes
fma.relu.{f16, f16x2} and fma{.relu}.{bf16, bf16x2} introduced in PTX ISA version 7.0.
Target ISA Notes
fma.relu.{f16, f16x2} and fma{.relu}.{bf16, bf16x2} require sm_80 or higher.
```

Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3938]: Oops good catch. Thanks

Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Fixed

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: I've added the PTX and arch requirements on the instruction, and the pattern Requires just on the pattern.

File: llvm/test/CodeGen/NVPTX/fma-relu.ll
Comment (hdelan) [reply llvm/test/CodeGen/NVPTX/fma-relu.ll:131]: I've added more tests to cover these cases.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3927]: Next question is -- what do we want to do about `.ftz`?
We handle it for regular FMA instructions and it's probably needed here, too.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (ldrumm) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3921]: It's not clear from the name that this is strictly *positive* zero. Maybe `def positive_zero_fp` or a better equivalent if you can

**Review by ldrumm — APPROVED** [review]
Looks good!

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3935]: The `relu` and `ftz` parts of the instructions appear to be in the order different from that in the PTX manual:

```
Syntax

fma.rnd{.ftz}.relu.f16      d, a, b, c;
fma.rnd{.ftz}.relu.f16x2    d, a, b, c;
```
Does ptxas accept generated instruction? In any case to avoid unnecessary divergence vs. the manual, it may be worth it to match the official syntax, even if ptxas happens to accept `fma.rn.relu.ftz.f16`

**Review by Artem-B — COMMENTED** [review]
We will eventually want to figure out a way to effectively use f16x2 and bf16x2 variants of `fma.rn.relu`.

**Comment by hdelan** [comment]
> Suppose the `fma` has more uses in addition to the `fmaxnum`, If this optimization kicks in it may increase the register pressure and won't be a clear win in terms of performance. I'm not sure this will be a problem, but to be conservative it may be better to implement this as a DAG combine and verify the `fma` has a single use.

I've changed the pattern matching to make sure FMA relu is only emitted if the FMA DAG has a single use.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: PTXAS did accept the instruction but I've updated to match the ordering specified here.

**Comment by hdelan** [comment]
> We will eventually want to figure out a way to effectively use f16x2 and bf16x2 variants of `fma.rn.relu`.

These should be available via the LLVM intrinsics. I think pattern matching to generate the x2 variants is a bit messy and fragile since it involves unpacking vector types. There also isn't really a concise way to express this pattern in source code, meaning the use cases for this sort of substitution are likely to be rare. Users needing this instruction for x2 types can always do so via the clang builtins.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Thanks have updated.

File: llvm/test/CodeGen/NVPTX/fma-relu.ll
Comment (hdelan) [reply llvm/test/CodeGen/NVPTX/fma-relu.ll]: This prevents `llc` from emitting PTX in some cases so I'm not going to add this flag.

File: llvm/test/CodeGen/NVPTX/fma-relu.ll
Comment (hdelan) [reply llvm/test/CodeGen/NVPTX/fma-relu.ll]: Thanks have changed.

**Comment by hdelan** [comment]
FTZ FMA is not supported for bf16, so I have removed some patterns elsewhere that were allowing `fma.rn.ftz.bf16` to be emitted.

File: llvm/test/CodeGen/NVPTX/fma-relu-contract.ll
Comment (Artem-B) [inline llvm/test/CodeGen/NVPTX/fma-relu-contract.ll:1413]: Another future optimization opportunity -- we could lower those "truncate-to-i16, then recombine into i32" as a single `prmt`

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3947]: How does FMA handle NaNs? `fmaxnum` requires `If either operand is a NaN, returns the other non-NaN operand.`

If we pass a NaN into an `fma.relu`, and get back a `NaN` back and that would be incorrect for `fmaxnum`.
The documentation does not seem to say anything about NaN handling by FMA (it only mentions it for .sat variants)

Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3947]: We are only applying this pattern if `allowUnsafeFPMath` is enabled. So AFAIA we don't need to worry about NaN propagation.

Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3947]: The comment for the `allowUnsafeFPMath` says that it only affects precision of the operations, not NaN handling:
https://github.com/llvm/llvm-project/blob/7f60f1312ae007d645fc96618db1238baea088b9/llvm/include/llvm/Target/TargetOptions.h#L171
So, it would allow using a regular FMA instead of `a*b+c`, but `NaN` propagation would not be affected.
This is not the case for `fmaxnum` -> `fma`, where `fmaxnum` has distinctly different requirements for processing NaN inputs.

In any case, we should run that by someone who's familiar with FP handling nuances. 
@arsenm @RKSimon -- what would be a sufficient constraint to reflect that the instruction is not expected to deal with NaNs?

Comment (arsenm) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3947]: Definitely do not rely on allowUnsafeFPMath for no nans. We really should delete allowUnsafeFPMath. 

FMA propagates nan as normal, and may also produce output nans based on the operands.

I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Thanks, I have changed the constraints to `Requires<[noNaNsFPMath]>`.

> I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

I have extended the fpimm pattern to also accommodate negative zero.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3908]: `fpimm_any_zero` ?

**Review by Artem-B — APPROVED** [review]
Nice. Please hold off on merging the patch till Monday, so if we find issues, there's someone around to deal with that.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Sure, have changed. Merging now.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: > I also think this pattern is broken for signed zeroes. I assume the ftz modifier is expected to preserve -0

I think I misunderstood what you were saying first time round. You are concerned with a comparison of `0.0` and `-0.0`. Since I am pattern matching with `fmaxnum`, which says either -0.0 or 0.0 may be returned in a comparison of both values, the guarantees of `fma.relu` (which doesn't specify what is returned when signed and unsigned zeros are in the final comparison) are the same as the `fmaxnum` comparison.

**Comment by hdelan** [comment]
I've also added extra cases so that `nnan .. fma` in the same pattern will allow the emission of `fma.relu`

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (frasercrmck) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3970]: I personally wouldn't bother supporting the function attribute case - I don't think it warrants the extra code and patterns. Supporting `nnan` should be enough. I think that the NVPTX backend should generally be moving towards using instruction flags rather than function attributes to do its various fast-math optimizations. But I'd defer to other people on that point.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (frasercrmck) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3926]: I'm sure a `multiclass` could whittle this bulk down quite a bit. The only things changing are the instruction suffix, the reg class, and the instruction suffix in the asm.

**Review by frasercrmck — COMMENTED** [review]
The PR description needs updating to reflect some of the new changes - the conditions under which the optimization is legal/performed, and that some instructions are being removed.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3936]: Nit: I'd prefer constructing the name up-front here, rather than deriving the final record name from the argument string.

E.g:
```
multiclass FMARELU<RegisterClass RC, string OpVecString = "">  {
  def _F16 : ...
  // FTZ variants are only supported by fp16, not bf16
  def _F16_FTZ : ...
  def _BF16: ...
}

defm FMARELU : FMARELU<Int16Regs>;
defm FMARELUv2 : FMARELU<Int32Regs, "x2">;
```

Comment (frasercrmck) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3936]: Sorry, that was my suggestion. I saw that `f16x2` and `bf16x2` were the more common names for these types in the instruction definitions, and thought it better to keep the instruction enums consistent at the expense of the TableGen. I can't see any existing examples of `v2_{type}`. There are a few examples of `v2{type}` but mostly for `b(8|16|32|64)` - not so much with specific types. In summary I suppose it's just a little inconsistent?

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (frasercrmck) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3908]: This might give a warning about bitwise orbeing used instead of logical or. I'd suggest using `||`.

Alternatively, perhaps `Imm.isExactlyValue(+0.0) | Imm.isExactlyValue(-0.0)` is just `Imm.isZero()`?

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3936]: I see. My concern was mostly this seemingly duplicate `defm FMARELU`, which is what got me to stop and poke at it. 

So, how can we express it all better. Ideally we want to have just one point to parametrize the multiclass.

We have 3 instruction variants, two vector sizes. From that we derive register size.
Considering that we're talking only about 6 items, there's not much point to do anything particularly complicated, as it would not improve anything over just enumerating things.

Actually, that may be the most concise way to do it -- just define a simple class for one instruction record that can parametrize ftz/type/regtype, and then define each of the 6 variants we have. You can see an example of this approach here:
https://github.com/llvm/llvm-project/blob/cb9481dbf902adc349757eca12a0a09396dc4a23/llvm/lib/Target/NVPTX/NVPTXInstrInfo.td#L1222

(side-note: FNEF_F16 and FNEG_BF16 could be further collapsed into a single class)

I think the net result will be shorter than what we have now.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: Have updated

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (arsenm) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: We are fixing the definition of fmaxnum/fminnum to require correctly ordered signed zeroes. The fuzzy behavior can be achieved with nsz, so it doesn't make sense to define it this way.

Best to respect this now to reduce the amount of code that needs changing

Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: I've added a PatFrag to make sure the `fmaxnum` has `nsz`.

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td]: I've updated to use `class` instead of `multiclass`

**Review by frasercrmck — APPROVED** [review]

File: llvm/lib/Target/NVPTX/NVPTXInstrInfo.td
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3970]: Nit: `OpVecString` should probably be called `Instruction` now.

Side note: It looks like we've simplified it to a pretty generic `3-operand instruction` helper class. We could probably use it in other places. It seems to be a fairly common pattern in other back-ends.

Comment (hdelan) [reply llvm/lib/Target/NVPTX/NVPTXInstrInfo.td:3970]: Thanks. Have updated. Since the commits are going to be squashed here I will leave the refactoring for a subsequent patch.

**Pull request merged**
